### PR TITLE
Windows Layout Improvements

### PR DESCRIPTION
**Pull request by cedricbahirwe**
- Improved App UI & Windows Sizes
- Removed unused code
- Fix Multiple navigation displaying on ExerciseListView when loading
- Optimized Light & Dark mode support

### CODE REVIEW

File: Exercism/AsyncResultView.swift [hunk 1]
```diff
@@ -11,25 +11,25 @@ struct AsyncResultView<Source: LoadableObject, Content: View>: View {
...
+        source: Source,
+        @ViewBuilder content: @escaping (Source.Value) -> Content
+    ) {
         self.source = source
         self.content = content
     }
     
     var body: some View {
-        Group {
+        ZStack {
```
Comment (angiemugo) [inline Exercism/AsyncResultView.swift:23]: I am curious why you changed this from a group?

File: Exercism/AsyncResultView.swift [hunk 2]
```diff
@@ -11,25 +11,25 @@ struct AsyncResultView<Source: LoadableObject, Content: View>: View {
...
             switch source.state {
             case .idle:
                 EmptyView()
             case .loading:
-                Spacer()
                 ProgressView()
-                Spacer()
+                    .frame(maxWidth: .infinity, maxHeight: .infinity)
```
Comment (angiemugo) [inline Exercism/AsyncResultView.swift:29]: great catch

File: Exercism/Common/ExercismButton.swift [hunk 3]
```diff
@@ -16,12 +16,16 @@ struct ExercismButton: View {
     var body: some View {
         ZStack(alignment: .center) {
             ProgressView()
+                .colorInvert()
```
Comment (angiemugo) [inline Exercism/Common/ExercismButton.swift:19]: great catch

File: Exercism/ExerciseList/ExerciseGridView.swift [hunk 4]
```diff
@@ -15,91 +15,120 @@ struct ExerciseGridView: View {
...
-            .onHover { hover in
-                if exercise.isUnlocked {
-                    withAnimation {
-                        isHover = hover
-                    }
+                    .font(.title3)
+            }
+            .frame(maxWidth: .infinity, alignment: .leading)
+            
+            Group {
```
Comment (angiemugo) [inline Exercism/ExerciseList/ExerciseGridView.swift:32]: why is this in a group?

File: Exercism/ExerciseList/ExerciseGridView.swift [hunk 5]
```diff
@@ -15,91 +15,120 @@ struct ExerciseGridView: View {
...
+                }
+                .padding(.vertical, 2)
+                .roundEdges(
+                    backgroundColor: Color.green.opacity(0.2), lineColor: .green)
+                .font(.callout.weight(.semibold))
+                .foregroundStyle(.green)
             }
             
-            if (solution?.numTterations ?? 0) > 0 {
+            if let numIterations = solution?.numTterations,
```
Comment (angiemugo) [inline Exercism/ExerciseList/ExerciseGridView.swift:74]: great catch

**Review by angiemugo — COMMENTED** [review]
Great job on this one. I left a few comments that are low priority

File: Exercism/ExerciseList/ExerciseGridView.swift (hunk 4 above)
Comment (cedricbahirwe) [reply Exercism/ExerciseList/ExerciseGridView.swift:32]: I wanted to apply a `.frame(maxHeight:)` to the resulting `Image` so it's centered vertically in the container, so I used the `Group` so the modifier could be applied without doubly applying it to both Images...
I guess we could also go with something like:
```
 (exercise.isUnlocked ? Image.chevronRight : Image.lock)
     .frame(maxHeight: .infinity)
```

**Comment by angiemugo** [comment]
One other thing, because the token in now saved in the keychain(which prompts a system login) could you test what happens when this login is unsuccessful? I am just seeing an empty screen

File: Exercism/AsyncResultView.swift (hunk 1 above)
Comment (cedricbahirwe) [reply Exercism/AsyncResultView.swift:23]: I was testing out whether the `Group` was causing some unexpected behaviour, I guess I forgot to change it back

**Comment by cedricbahirwe** [comment]
I guess, I need to revert to this, because my new implementation of that might not be accurate
Screenshot 2024-02-19 at 10 20 53

**Comment by angiemugo** [comment]
Yes @cedricbahirwe , this works now. What is  happening is if user is logged in there is no destination specified.

**Comment by cedricbahirwe** [comment]
> Yes @cedricbahirwe , this works now. What is happening is if user is logged in there is no destination specified.

Yeah, but in the logic, shouldn't `performInitialNavigationSetup` get the saved path where the view appears and assign it to the `navigationModel`?

The reason I previously removed the code in the picture, is that, `TracksListView` was called twice (triggering multiple API calls than necessary...

**Comment by angiemugo** [comment]
> > Yes @cedricbahirwe , this works now. What is happening is if user is logged in there is no destination specified.
> 
> Yeah, but in the logic, shouldn't `performInitialNavigationSetup` get the saved path when the view appears and assign it to the `navigationModel`?
> 
> The reason I previously removed the code in the picture, is that, `TracksListView` was called twice (triggering multiple API calls than necessary...

Yes you're right. I'll investigate where this logic is failing and push a fix

**Comment by cedricbahirwe** [comment]
Hey @angiemugo , I might have been wrong on some of my previous observations, It looks like when a window is closed on macOs, `SceneStorage` data might also be destroyed, which I did not account for during refactoring:

Screenshot 2024-02-23 at 22 29 10


In our previous logic, we have the logic below, which means that if the keychain token exists (user logged in) even though `SceneStorage` value returns `nil`, the app would still show the `TracksListView`.
```swift
if let _ = ExercismKeychain.shared.get(for: Keys.token.rawValue) {
    TracksListView(
        asyncModel: .init { try await model.getTracks() }
    )
    .frame(minWidth: 800, minHeight: 800)
    .environmentObject(navigationModel)
    
} else {
    LoginView().frame(minWidth: 800, minHeight: 800)
}
```

To handle that case, I added the logic below:

```swift
if let navigationData {
    navigationModel.jsonData = navigationData
}

// User is logged in but navigationModel path is empty
if ExercismKeychain.shared.get(for: Keys.token.rawValue) != nil && navigationModel.path.isEmpty {
    navigationModel.goToDashboard()
    // Save navigation data into scene storage
    navigationData = navigationModel.jsonData
}

for await _ in navigationModel.objectWillChangeSequence {
    navigationData = navigationModel.jsonData
}
```

I've pushed the code in this PR, you can try it out and let me know

**Comment by angiemugo** [comment]
I'd also like for you to look at the Xcode cloud workflow. I already erroneously merged it into main.  lmk what you think and what addition we can make

**Review by angiemugo — APPROVED** [review]

**Comment by cedricbahirwe** [comment]
Can we merge this or shall we wait @angiemugo ?

**Comment by angiemugo** [comment]
Yeah, let's merge it

**Pull request merged**